src/controller/CHIPDevice.h

```diff
@@ -482,6 +492,9 @@ class DLL_EXPORT Device : public Messaging::ExchangeDelegate, public SessionEsta
 
     uint8_t mCSRNonce[kOpCSRNonceLength];
 
+    uint8_t mNOC[Credentials::kMaxDERCertLength];
```

Comment:
This is invalid: where this is used, the NOC is a chain received by the CA, and it should be in TLV certificate format, and must allow for an ICAC int he chain (which this array is too small to support).

Where the call:
```
        err = commissioner->GenerateOperationalCertificates(device->GetNOC(), chipCertSpan);
        SuccessOrExit(err);
``` 

is made, looking at the logic, it sounds like:

1. Ask the device for its NOC chain
2. Generate a NOC (from a NOC?)
3. Send the NOC

There are several concepts that are mixed here and that do not match the flow's intent.

The "device" object here is used as a bucket to receive lots of temporary data from the commissioning flow, where it could instead be maintained in a commissioning context.

The flow also prevents usage of an ICAC in the chain.

I understand not all of it can be refactored at once, but here we are adding new code that does very different than intended.

At the very least we should not mix "NOC" terminology between a given commissioner's preferred internal format (which may be DER) and the actual need of the protocol: that it be TLV.

I would recommend:

1. Allocating enough space for the whole chain of NOC in matter certificate format root + ICAC + RCAC
2. When the call to the CA is made to go from NOCSR to NOC chain, keep all consideration of DER format purely internal, and output root cert and ICAC+NOC separately, in Matter format
3. Here, rename mNOC to mNOCChain, and have enough space for ICAC+NOC
4. Here, add storage for mRootCert --> Note that this is not multi-admin safe, but will do for now...
5. After obtaining the operational chain from the processing of the NOCSR, keep it "handy" while the root is being installed.

In other words, the following code:

```
    {
        MutableByteSpan chipCertSpan(chipCert.Get(), chipCertAllocatedLen);
        err = commissioner->GenerateOperationalCertificates(device->GetNOC(), chipCertSpan);
        SuccessOrExit(err);

        ChipLogProgress(Controller, "Sending operational certificate to the device");
        err = commissioner->SendOperationalCertificate(device, chipCertSpan);
        SuccessOrExit(err);
    }
```

Should only be sending the NOC (second step), rather than converting.

Reply:
@tcarmelveilleux, a few questions..

CA Interface (`OperationalCredentialsDelegate`)
1. The generated certificates are in DER format. I think that's the right approach, as it keeps matter specific logic (conversion to CHIPCert) contained within common part of the code. Each delegate implementation doesn't have to convert DER certs to CHIPCert before returning via the delegate API.
2. Currently, there are 3 APIs to return root cert, ICA cert and operational cert, respectively. Does it make sense to condense it to a smaller set? I think, keeping root cert API as a separate call makes sense, but ICA and opcert can definitely be combined into one.
3. Make all the APIs async. Right now only opcert API is async. 

Currently, when the commissioner first starts, it gets its own credentials using the CA Interface (`OperationalCredentialsDelegate`). The device is provisioned with the root cert that the commissioner was assigned. If that's the right thing to do, we may not need to store the root cert in the `device` objects (the `device` object is a proxy object in controller/commissioner for the actual device).

Similarly, we don't need to store the ICAC in the `device` object, as we send it to the actual device immediately after we get it from the CA. If we agree to condense the opcert, and ICAC APIs in CA Interface (`OperationalCredentialsDelegate`), it'll be required (and ok) to store both certs in `device` object.

Lastly, we currently have API (`ConvertX509CertsToChipCertArray`) to convert two x509 certs (ICAC and OpCert) to CHIPCert array. Storing the opcert temporarily in x509 format helps, as we can call `ConvertX509CertsToChipCertArray` to combine the stored opcert, and received ICAC into on CHIPCert array. If we agree to condense opcert and ICAC APIs into one API, this part of the discussion can be ignored, as we can just convert the received certs to CHIPCert and store that in `device` object.

Agree on renaming code to match the spec, and disambiguate it.